### ISSUE TITLE
zero confirmation spend - fees incorrect? build_block/process_block mismatch?

### ISSUE THREAD

**Issue by antiochp** [ignopeverell/grin#181]
During some local testing of zero confirmation spends from a wallet, I am seeing outputs generated that never appear to get confirmed (so suspect the values are incorrect in the wallet).

For example - 

```
Spent 1000000000 
To send 5000000 
Received 4999920 (fees 80)
Change output 995000000
```

```
Spent 995000000 (zero confirmation from above)
To send 5000000 
Received 4999920 (fees 80) <-- never gets confirmed?
Change output 990000000 <-- never gets confirmed?
```

We see this in node logs -

```
Oct 15 20:18:27.867 DEBG block_fees here: BlockFees { fees: 80, height: 6, key_id: Some(Identifier(f26dbc24171432f88888)) }
Oct 15 20:18:27.994 DEBG (Server ID: Port 13414) Built new block with 1 inputs and 3 outputs, difficulty: 10
```
This is the _only_ block in the logs with inputs, so both txs are presumably in there (one spending the change output of the other).

I think the fees on this block are being calculated incorrectly.
There should be 80 in fees _for each_ tx here, 160 in total?

`wallet info` looks like - 

```
Outputs - 
key_id, height, lock_height, status, zero_ok, value
----------------------------------
17ebdeb665384ae5db17, 1, 4, Spent, false, 1000000000
a9d252771e29beb5cce8, 2, 5, Unspent, false, 1000000000
dd930ae6719a382182b2, 3, 6, Unspent, false, 1000000000
4636dd289d8f23d93a0c, 4, 7, Unspent, false, 1000000000
3f9002bfecc722ade9bc, 5, 8, Unspent, false, 1000000000
9b06dbf15475d98a65bf, 0, 0, Spent, true, 995000000
06c43cafdb59ae893b4e, 6, 0, Unspent, false, 4999920
86767a2b6d44795db540, 0, 0, Unconfirmed, true, 990000000
44a325a40d2c726bf074, 0, 0, Unconfirmed, false, 4999920
f26dbc24171432f88888, 6, 9, Unspent, false, 1000000040
4b443bf73a62d814d4b2, 7, 10, Immature, false, 1000000000
7f7d3f5ea9dcf0e1b0ee, 8, 11, Immature, false, 1000000000
b41e22cd4204c0454381, 9, 12, Immature, false, 1000000000
268081b71c66134a8c91, 0, 0, Unconfirmed, false, 1000000000
```

**Comment by antiochp** [comment]
Digging in further it looks like `get_mineable_transactions` only returns the `roots`.
I was initially thinking this meant we only took those txs into account for the fee but now I'm thinking the child txs are not actually being included in the block at all?
But they *are* being removed from the pool?

**Comment by antiochp** [comment]
If I update `get_mineable_transactions` to naively return both `roots` and all `vertices`
then I see the expected behavior - 
* the outputs of the 2nd tx appear to be confirmed and become spendable in the wallet.
* the fees on the block that includes both txs is 160

```
Oct 16 10:31:56.896 DEBG block_fees here: BlockFees { fees: 160, height: 6, key_id: Some(Identifier(f26dbc24171432f88888)) }
Oct 16 10:31:57.096 DEBG (Server ID: Port 13414) Built new block with 1 inputs and 4 outputs, difficulty: 10
```

I *think* 1 input and 4 outputs is correct here - 
- 1 coinbase output.
- 1 original input creates 2 outputs.
- 2nd input cancels out of of these outputs and creates 2 new outputs.

**Comment by antiochp** [comment]
The miner calls the following two functions -

```
build_block 
    -> prepare_mineable_transactions 
        -> get_mineable_transactions

process_block
    -> block_accepted 
        -> reconcile_block 
            -> "mark and sweep algo"
```


I'm not familiar with the intended process here but it seems like it is critical we keep the list of txs consistent across both of these.
i.e. whatever txs end up in the block being mined **must** be the same as the list of txs that get removed from the pool via the mark and sweep.

* If we mine a tx that does not get included in the mark and sweep then it ends up as a double spend on the next block getting mined (the miner will attempt to include it a 2nd time).
* If we mark and sweep a tx that does not get included in the block - then we end up in the situation above and txs get dropped without ever getting mined.

It feels like we should be doing the "mark" portion of this *before* we choose the txs to be included in the current block. i.e. we mine all the "marked" txs.
And then we subsequently "sweep" that same list of txs to keep the pool correctly sync'd.

@ignopeverell - I'm sure I'm missing something here so what's wrong with my understanding of the above?

**Comment by MoaningMyrtle** [comment]
Hi there,

Gonna take a stab at clarifying/answering a few things.

First off, selecting just the roots is intended right now and is definitely something that should be improved. It's the minimum viable miner that works, is consistent, and selects transactions properly without errors- but obviously its not ideal, since it doesn't select transactions in a way that enables block cut-through, which is pretty critical for scalability and privacy reasons.

At the time I opted not to implement anything further as getting the minimum viable miner working seemed like a reasonable goal that didn't swallow too much complexity. It seemed reasonable to me to start a testnet with as straightforward of a system as possible before making the pool more complicated, as its already a pretty tricky system to reason about. It may be time to rectify this, I'll defer to @ignopeverell about that.

The consequence of this is that spending an unconfirmed in the current system will take at least 2 blocks to confirm both transactions.

As a stop-gap before true transaction selection (which would likely involve things like fee/priority, computing cut-through and impacts on size, etc), returning the entire pool, traversed from the roots up, would produce a valid block (as long as the size of the resulting transaction does not exceed limits).

>it seems like it is critical we keep the list of txs consistent across both of these.
> i.e. whatever txs end up in the block being mined must be the same as the list of txs that get removed from the pool via the mark and sweep.

From a high level, this is true, but in practice it's impossible to enforce this on a distributed system. Nodes will diverge, their pools will have different transactions (sometimes double spends) and their contents are not consensus-enforced. The end result is that I chose to go for a best-effort approach with a focus on consistency- that is,  when building a block we should return only mineable transactions (transactions rooted to the blockchain or another mineable transaction included in the same block), and when rectifying a block, the end result of the pool should only contain transactions which are mineable by the consensus rules known at the time (no duplicate output commitments, no double-spends, etc.)

Furthermore, we cannot guarantee (and should not rely on the assumption that) a miner is using the same pool implementation as this one. The only "correct" thing to do is to remove transactions based purely on what the incoming block says, not what this node prefers- which is what the existing `reconcile_block` mechanism does. 

The guarantee provided by the existing transaction pool is that:

1. All transactions returned during block building can be included in the next block, to the best understanding of this node of the global consensus state.

2. After a block is received, all transactions remaining in the pool should be consistent with the blockchain's state with respect to consensus rules.

> If we mine a tx that does not get included in the mark and sweep then it ends up as a double spend on the next block getting mined (the miner will attempt to include it a 2nd time).

This would definitely be a bug. Are you saying you've observed this in practice? The existing system should not allow this.

> If we mark and sweep a tx that does not get included in the block - then we end up in the situation above and txs get dropped without ever getting mined.

While dropping transactions that have not yet been mined is to an extent unavoidable (consider the max pool size/expiration case), we should definitely try to minimize the occurrence. Have you observed a transaction being eliminated that shouldn't be?

> "mark" process is the same as choosing txs to mine)

This is not quite right, as implemented now. "Mark" looks at the outputs created and consumed in a block and only eliminates those transactions that are invalidated by the block. This may include transactions not included in the block, if it would result in a violation of consensus rules to include them. It does not consider this node's transaction preference, or any such preference (including fee-based), only what is inside the block.

On the other hand, transaction selection utilizes the node's transaction preference, which may or may not include fee-based and other parameters, to determine which transactions it _wants_ to include.

**Comment by antiochp** [comment]
Hey @MoaningMyrtle - thanks for the info. This is really useful.

> The consequence of this is that spending an unconfirmed in the current system will take at least 2 blocks to confirm both transactions.

This does not appear to be what I am observing in the current system.
If a tx is not included in the first block it is not included in any blocks as far as I have observed.

> While dropping transactions that have not yet been mined is to an extent unavoidable (consider the max pool size/expiration case), we should definitely try to minimize the occurrence. Have you observed a transaction being eliminated that shouldn't be?

Yes I _think_ so (but granted I'm not 100% sure as this is pretty complex to debug).
Every time we call `reconcile_block` it recursively removes all txs from the pool based on the incoming block (v.s. `build_block` which deals purely with the top level `roots`).
These will always differ for a zero confirmation tx and I think I am observing non-root txs consistently being dropped.

From the original example in this issue - the outputs of the 2nd tx are being removed from the pool.
They are not in the `roots` of the pool so never get mined in the first block.
And the subsequent blocks do not include them as they are no longer in the pool.

I agree that for simplicity in an initial testnet it may make sense to only include the roots when building a block.
But - we would need to ensure that everything else remains in the pool for inclusion in subsequent blocks. In which case (at least intuitively) the "mark and sweep" would simply looks for roots and remove those (and would not need to be recursive at all)?

**Comment by antiochp** [comment]
@MoaningMyrtle - I just pushed a failing test case in #182 to illustrate this.

**Comment by antiochp** [comment]
To restate my understanding of the issue here (please correct any misunderstandings) - 

* the pool is effectively a set of directed acyclic graphs of txs from the roots
* a miner is free to select whatever txs they want to include in the next block
  * tx selection is not consensus enforced, but - 
  * the block itself *must* be valid (this *is* consensus enforced)
* reconciling the pool needs to result in a pool of txs that are all valid (consensus enforced)
  * spent outputs must be evicted from the pool (avoiding double spend)
  * we should aim to minimize evicting still valid txs from the pool (they will be lost forever)

But - 
* for any given input to a block, the miner could have selected any arbitrary subgraph of vertices from the pool (while still producing a valid block) 
* and given the possibility of cut-through within the block

Then it would appear impossible for another node to determine with any level of certainty which txs to evict from the pool when reconciling a block.
It cannot traverse the graphs in the pool based on the inputs to the block as the subgraphs of vertices originally selected for inclusion in the block are unknown.

**Comment by MoaningMyrtle** [comment]
Hi there,

Going to dig more and double check the test execution to make sure there hasn't been a regression, but your test case appears to be missing a call to `chain_ref.apply_block(&block);` before block reconciliation. The pool expects the chainstate to be updated before entry to the `reconcile_block` method. The existing `test_block_reconciliation` has pretty good coverage of these basic cases. It currently exercises the same situation your test attempts to, although without the use of the keychain/signing process. The transaction `pool_child` from `test_block_reconciliation` is a transaction spending a pool transaction, which is verified to be still a part of the pool after block reconciliation- the output `3` is checked against the pool in https://github.com/ignopeverell/grin/blob/8f76746e84c1fb26e3eb5aec1ad73707c67897fa/pool/src/pool.rs#L1014.

In a nutshell that also answers your question here:

> It cannot traverse the graphs in the pool based on the inputs to the block as the subgraphs of vertices originally selected for inclusion in the block are unknown.

This is absolutely correct, but that doesn't preclude the pool from generating correct output.  It may be somewhat unintuitive, but as long as the pool removes anything that does conflict, and minimizes the removal of things that don't conflict, it doesn't matter if the transactions removed are the same as those included in the block. (For example, in the case of a conflicting double spend in the pools of different nodes, that's the _only_ thing you can do. There is no way to remove the same transaction).

What the pool does is utilizes the block to understand what outputs are consumed by the block's inputs, but it goes back to the consensus chainstate to independently verify and connect each of the newly disconnected outputs back to the chain, evicting any that cannot. So, in the case that an output included in a block has been cut-through and can't be determined by the receiving node's pool, the end result is it doesn't matter- if the block spends the output of a parent pool transaction, and the new chainstate doesn't provide the output needed to satisfy the newly-disconnected child, that child is removed, regardless of the root cause of the removal.

Think of the process in `reconcile_block` as follows:

1. Prior to reconciliation, we have a graph where all pool transactions' spending references (inputs) are connected to other pool transactions or to the blockchain.
2. When the block is received, it consumes some outputs (the block's inputs). It also produces some outputs, which should be added to the chain's consensus chainstate.
3. Some transactions spent outputs that have been consumed by a block. These outputs are no longer valid for pool transactions to spend. The reason why they are no longer valid isn't important. It could be that the pool transaction is exactly spent in the block, but it may just be a conflicting confirmed transaction- either way, that pool transaction needs to be evicted.
4. The `Mark` step examines these conflicting transactions, then walks the graph down to their dependent children. For each child, it checks that the missing output(s) are satisfied by the chain's newly-updated consensus chainstate. Based on a previous invariant (no duplicate outputs), we know that the new output cannot be spent by another pool transaction in addition to the child we're currently looking at, so if the blockchain utxo set provides the output we need we know the transaction is still valid. If it doesn't, the child transaction is also marked for eviction, and the system recurses deeper.
5. The `Sweep` step then cleans up the transactions that have been marked for eviction, replacing the references (vertices) with new ones that correspond to the updated chainstate.

Another way to look at it is the pool doesn't treat the block as anything special- it behaves as if the block is one giant transaction, which takes priority over existing transactions in the pool, and then does what it needs to do to clean the pool up after accepting that meta-transaction. This is necessary because of the possibility of cut-through, which means we cannot rely on the ability to reverse transactions from the block for the purposes of eviction (a conclusion you've already reached).

Definitely not implying that the algorithm is foolproof- but the test in https://github.com/ignopeverell/grin/blob/8f76746e84c1fb26e3eb5aec1ad73707c67897fa/pool/src/pool.rs#L873 covers a bunch of weird, unintuitive transaction patterns and behaviors  (including double spends and transactions that behave like cut-throughs) that might help clarify what is and isn't a property of the pool and block reconciliation at this time.

**Comment by antiochp** [comment]
Mind blown - the part in (4) about the consensus chainstate is what I was missing. Thanks for the explanation.
Adding the missing call to `chain_ref.apply_block(&block);` gets the test to pass.

But - this does not appear to be what we are seeing in practice with a node running against a real blockchain impl (and not the dummy chain).
So I'm continuing to dig in deeper - seems like one possibility is a difference between the behavior in `extension.apply_block()` and `DummyChain.apply_block()`?

If this turns out to indeed be a non-issue then apologies... - but I've definitely learnt a lot from this exploration.

**Comment by antiochp** [comment]
@MoaningMyrtle @ignopeverell 
Found the issue here - 

We push 2 txs to the pool.
When we call `get_mineable_transactions` for the next block we see the pool looks like this - 
```
roots: [9b52b108] # one root tx
vertices: [319ac99c] # one non-root tx
```
Then we successfully mine that block including tx `9b52b108`.

Now when we call `get_mineable_transactions` for the next block we see the pool now looks like this - 
```
roots: [] # ***no root txs***
vertices: [319ac99c] # one non-root tx
```

So we _have_ successfully evicted the tx `9b52b108` from the pool in `reconcile_block` and left the tx `319ac99c` in the pool. 

__But__ - we have not updated the list of root txs in the pool.

When we select txs for the next block we do not select `319ac99c` as it is not a root tx in the pool.
It's in there - we just never select it.

**Comment by antiochp** [comment]
Updated the test in #182 to illustrate this.

**Comment by MoaningMyrtle** [comment]
Awesome work catching that! I'll take a look at your patch and comment over on #182

**Comment by antiochp** [comment]
Resolved in #182
